{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c47602df-c6a1-4f68-b946-585d0f90ce3e",
   "metadata": {
    "tags": []
   },
   "source": [
    "# How to use Cartography\n",
    "This is a pre-alpha version of Cartography for use by the Annotation team.  \n",
    "It's broken and buggy in lots of places, but we're working on making it more robust!  \n",
    "\n",
    "These instructions are for running the code in a Jupyter lab environment that has already been set up to run Cartography.  \n",
    "If you want to set up your own environment locally, see **[Local Setup](#Local-Setup)**.\n",
    "\n",
    "---\n",
    "## Vanilla Mode\n",
    "Starts with input FASTA proteins, performs FoldSeek and BLAST, and builds a space.\n",
    "\n",
    "![vanilla rulegraph](rulegraph.png)\n",
    "\n",
    "#### 1. Open up a new terminal.  \n",
    "- Click the blue \"+\" button on the top left.  \n",
    "- Under \"Other\", click \"Terminal\".  \n",
    "\n",
    "#### 2. Activate the conda environment.  \n",
    "- Run this command in the Terminal:  \n",
    "    `conda activate cartography`\n",
    "    \n",
    "#### 3. Add query files to the `input/` folder.\n",
    "- The pipeline will search these proteins against the BLAST NR database and the Foldseek afdb50, afdb-proteome, and afdb-swissprot.\n",
    "- You can manually add FASTA files less than 400aa long, one file per peptide. The pipeline runs ESMFold to generate a predicted structure.\n",
    "- For longer sequences, use [CoLabFold](https://colab.research.google.com/github/sokrypton/ColabFold/blob/main/AlphaFold2.ipynb).\n",
    "- You can also pull down proteins with a Uniprot accession number and Alphafold structure using the following command.  \n",
    "    Replace `{accession}` with your Uniprot accession (e.g. [`P24928`](https://www.uniprot.org/uniprotkb/P24928/entry)).  \n",
    "    \n",
    "    `python ProteinCartography/fetch_accession.py -a {accession} -o input -f fasta pdb`\n",
    "    \n",
    "    - This saves a FASTA file from Uniprot and a PDB file from AlphaFold to the `input/` folder/\n",
    "- Sequences >700aa in length currently cause memory failures for remote blast.\n",
    "\n",
    "#### 4. Run the pipeline.\n",
    "- Run this command in the Terminal:  \n",
    "    `snakemake --cores 16`\n",
    "- You can also give a nickname to the analysis using:\n",
    "    `snakemake --cores 16 --config analysis_name={my_name}`\n",
    "\n",
    "---\n",
    "## From-Folder Mode\n",
    "Starts with a folder of PDB files and a `uniprot_features.tsv` file and builds a space.\n",
    "\n",
    "![from folder rulegraph](rulegraph_ff.png)\n",
    "\n",
    "#### 1. Put all the PDB files in a folder\n",
    "- You can do this on your personal computer.\n",
    "- Each protein should have a unique ID as the filename prefix, referred to as the \"protid\". For example, a protein with the protid `P00000` should have a PDB file with the name `P00000.pdb`.\n",
    "\n",
    "#### 2. Make a `uniprot_features.tsv` file.\n",
    "- For each protid in the dataset, you should gather features from Uniprot into a `uniprot_features.tsv` file.\n",
    "- The first column of the file should be the `protid`, e.g. `P00000` above, with `protid` as the column name.\n",
    "- You should also have additional columns with the following values for each protein:\n",
    "\n",
    "| column name | data type | description |\n",
    "| :---------- | :-------- | :---------- |\n",
    "| `proteinDescription.recommendedName.fullName.value` | `str` | a short human-readable description of the protein |\n",
    "| `organism.commonName` | `str` | the common name of the organism |\n",
    "| `organism.scientificName` | `str` | the scientific name of the protein |\n",
    "| `sequence.length` | `str` | the number of amino acids in the protein |\n",
    "| `annotationScore` | `int` | the [Uniprot Annotation Score](https://www.uniprot.org/help/annotation_score) for the protein. Should be 1 if it's a new protein. |\n",
    "| `organism.lineage` | `list` of `str` | the taxonomic lineage of the organism. each taxonomic group associated with the protein should be a string (in quotation marks) separated by a comma. Square brackets should surround the comma-separated list so that it is converted to a Python list upon reading with the [`eval()`](https://docs.python.org/3/library/functions.html#eval) built-in function. |\n",
    "\n",
    "> For `organism.lineage`, an example for a mouse protein might look like this:  \n",
    "> `['Eukaryota', 'Metazoa', 'Chordata', 'Craniata', 'Vertebrata', 'Euteleostomi', 'Mammalia', 'Eutheria', 'Euarchontoglires', 'Glires', 'Rodentia', 'Myomorpha', 'Muroidea', 'Muridae', 'Murinae', 'Mus', 'Mus']`\n",
    "\n",
    "- You should add this file to the same folder as the PDBs.\n",
    "\n",
    "#### 3. Upload the files to this compute environment.\n",
    "- You might want to zip the folder.\n",
    "\n",
    "#### 4. Update the `config_from_folder.yml` file.\n",
    "- Change the value of `input_dir` to be the folder of the PDBs that you uploaded.\n",
    "\n",
    "#### 5. Activate the conda environment.  \n",
    "- Run this command in the Terminal:  \n",
    "    `conda activate cartography`\n",
    "\n",
    "#### 6. Run the pipeline.\n",
    "- Run this command in the Terminal:  \n",
    "    `snakemake --snakefile Snakefile_from_folder --cores 16`\n",
    "- You can also give a nickname to the analysis using:\n",
    "    `snakemake --snakefile Snakefile_from_folder --cores 16 --config analysis_name={my_name}`\n",
    "\n",
    "---\n",
    "## Local Setup\n",
    "\n",
    "If you'd like to run Cartography on your local computer instead, you can try cloning the repo using GitHub and setting up the Conda environment instead.\n",
    "\n",
    "#### 1. Clone from GitHub\n",
    "- `git clone https://github.com/Arcadia-Science/gene-family-cartography.git`\n",
    "\n",
    "#### 2. Setup conda environment\n",
    "- `cd` into the github repo (e.g. `cd gene-family-cartography`)\n",
    "- `conda env create -n cartography -f envs/cartography.yml`\n",
    "- This can be slow, so you could install `mamba` first and then run the above conda creation using `mamba env create` instead.\n",
    "    - `conda install mamba`\n",
    "    - `mamba env create -n cartography -f envs/cartography.yml`\n",
    "\n",
    "#### 3. Get Started\n",
    "- See **[Vanilla Mode](#Vanilla-Mode)** or **[From-Folder Mode](#From-Folder-Mode)** above for your use case.\n",
    "\n",
    "--- \n",
    "## Known Bugs\n",
    "\n",
    "#### Foldseek Connection Failure\n",
    "The first time we run the pipeline, we often get an error from the rule `run_foldseek`:  \n",
    "`requests.exceptions.ConnectionError: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))`\n",
    "\n",
    "This has something to do with the speed of our requests.  \n",
    "Cancelling the pipeline with <kbd>Control</kbd> + <kbd>C</kbd> and then rerunning the `snakemake --cores 16` command seems to get around this.  \n",
    "\n",
    "#### BLAST dies\n",
    "Sometimes BLAST will take forever to respond or will die with an error like `CPU usage limit was exceeded, resulting in SIGXCPU (24)`.  \n",
    "You can wait a few minutes and try again later. Make sure to delete the `blastresults/` folder in `output/` so Snakemake knows to re-run that part of the pipeline.\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2fedee9d-9236-4b25-b4d3-6c51b6a135c6",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b8af3c6a-cf8a-4feb-88bc-0828ed930a25",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
